From: 1be95325d320122efd5dedf7437839cfcca01f7a
From: https://github.com/void-linux/void-packages/commit/1be95325d320122efd5dedf7437839cfcca01f7a
Date: Mon, 23 Sep 2024
Subject: [PATCH] onboard: fix segfault when hovering over the keyboard

Currently, if the mouse pointer is hovered over the Onboard keyboard, the application crashes with a segmentation fault.
This is because the new_device_event function is not acquiring the GIL before creating a new object. This patch fixes
the issue by acquiring the GIL before creating a new object. It also acquires the GIL before queueing the event.
This patch is taken from the fix provided in the following link:
https://github.com/void-linux/void-packages/commit/1be95325d320122efd5dedf7437839cfcca01f7a

Signed-off by: Pratheeksha S N <pratheeksha.s.n@ni.com>

Upstream-Status: Not Submitted [This is being added to the meta-nilrt layer for now. If it is accepted upstream, it will be removed from here.]

--- a/Onboard/osk/osk_devices.c
+++ b/Onboard/osk/osk_devices.c
@@ -97,13 +97,15 @@ osk_device_event_dealloc (OskDeviceEvent
 static OskDeviceEvent*
 new_device_event (void)
 {
-    OskDeviceEvent *ev = PyObject_New(OskDeviceEvent, &osk_device_event_type);
+    OskDeviceEvent *ev;
+    PyGILState_STATE gstate = PyGILState_Ensure();
+    ev = PyObject_New(OskDeviceEvent, &osk_device_event_type);
     if (ev)
     {
         osk_device_event_type.tp_init((PyObject*) ev, NULL, NULL);
-        return ev;
     }
-    return NULL;
+    PyGILState_Release(gstate);
+    return ev;
 }
 
 static PyObject *
@@ -334,6 +336,7 @@ osk_devices_dealloc (OskDevices *dev)
 static void
 queue_event (OskDevices* dev, OskDeviceEvent* event, Bool discard_pending)
 {
+    PyGILState_STATE state = PyGILState_Ensure ();
     GQueue* queue = dev->event_queue;
     if (queue)
     {
@@ -364,6 +367,7 @@ queue_event (OskDevices* dev, OskDeviceE
         Py_INCREF(event);
         g_queue_push_head(queue, event);
     }
+    PyGILState_Release (state);
 }
 
 static gboolean idle_process_event_queue (OskDevices* dev)
